File: Note Taking Application/backend/src/dao/NoteDAO.java
package dao;
import models.Note;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Timestamp;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
public class NoteDAO{
    Note note;
    List<Note> notes; 
    public boolean createNote(Note note){
          String sql = "INSERT INTO notes (title, content, dateCreated, modifiedDate, historyID, userID) VALUES (?, ?, ?, ?, ?, ?)";

        try (Connection conn = DBConnection.getConnection();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setString(1, note.getTitle());
            stmt.setString(2, note.getContent());
            stmt.setTimestamp(3, Timestamp.valueOf(note.getDateCreated()));

            if (note.getModifiedDate() != null) {
                stmt.setTimestamp(4, Timestamp.valueOf(note.getModifiedDate()));
            } else {
                stmt.setNull(4, java.sql.Types.TIMESTAMP);
            }

            stmt.setInt(5, note.getHistoryID());
            stmt.setInt(6, note.getUserID());

            int rowsAffected = stmt.executeUpdate();
            return rowsAffected > 0;

        } catch (SQLException e) {
            e.printStackTrace();
            return false;
        }
    }
    public List<Note> getallNotes(int userID){
        notes = new ArrayList<>();
        String sql = "SELECT * FROM notes WHERE userID = ?";
        try (Connection con = DBConnection.getConnection();
            PreparedStatement stmt = con.prepareStatement(sql)){
                stmt.setInt(1,userID);
                ResultSet rs = stmt.executeQuery();
                while (rs.next()){
                int noteId = rs.getInt("noteId");
                String title = rs.getString("title");
                String content = rs.getString("content");

                Timestamp createdTs = rs.getTimestamp("dateCreated");
                Timestamp modifiedTs = rs.getTimestamp("modifiedDate");

                LocalDateTime dateCreated = createdTs.toLocalDateTime();
                LocalDateTime modifiedDate = (modifiedTs != null) ? modifiedTs.toLocalDateTime() : null;

                int historyID = rs.getInt("historyID");
                int dbUserID = rs.getInt("userID");

                Note note = new Note(noteId, title, content, dateCreated, modifiedDate, historyID, dbUserID);
                notes.add(note);
                }
            }
            catch(SQLException e){
                e.printStackTrace();
            }
            return notes;
            }
    
    public Note getNotebyid (int noteID, int userID){
        String sql = "SELECT * FROM notes WHERE noteId = ?";
        Note note1 = new Note(userID);
        try (Connection con = DBConnection.getConnection();
            PreparedStatement stmt = con.prepareStatement(sql)){
                stmt.setInt(1,noteID);
                ResultSet rs = stmt.executeQuery();
                while(rs.next()){
                    int noteId = rs.getInt("noteId");
                String title = rs.getString("title");
                String content = rs.getString("content");

                Timestamp createdTs = rs.getTimestamp("dateCreated");
                Timestamp modifiedTs = rs.getTimestamp("modifiedDate");

                LocalDateTime dateCreated = createdTs.toLocalDateTime();
                LocalDateTime modifiedDate = (modifiedTs != null) ? modifiedTs.toLocalDateTime() : null;

                int historyID = rs.getInt("historyID");
                int dbUserID = rs.getInt("userID");

                note1 = new Note(noteId, title, content, dateCreated, modifiedDate, historyID, dbUserID);
                }
            }
            catch(SQLException e){
                e.printStackTrace();
            }
            return note1;
    }
    public boolean updateNote(Note note){
        String sql = "UPDATE Note SET title = ?, content = ?, modifiedDate = ?, historyID = ?, userID = ? WHERE noteID = ?";

    try (Connection conn = DBConnection.getConnection();
         PreparedStatement stmt = conn.prepareStatement(sql)) {

        stmt.setString(1, note.getTitle());
        stmt.setString(2, note.getContent());

        if (note.getModifiedDate() != null) {
            stmt.setTimestamp(3, Timestamp.valueOf(note.getModifiedDate()));
        } else {
            stmt.setNull(3, java.sql.Types.TIMESTAMP);
        }

        stmt.setInt(4, note.getHistoryID());
        stmt.setInt(5, note.getUserID());
        stmt.setInt(6, note.getNoteID());

        int rowsAffected = stmt.executeUpdate();
        return rowsAffected > 0;

    } catch (SQLException e) {
        e.printStackTrace();
        return false;
    }
    }
    public boolean deleteNote(int noteID){
       String sql = "DELETE FROM Note WHERE noteID = ?";

    try (Connection conn = DBConnection.getConnection();
         PreparedStatement stmt = conn.prepareStatement(sql)) {

        stmt.setInt(1, noteID);

        int rowsAffected = stmt.executeUpdate();
        return rowsAffected > 0;

    } catch (SQLException e) {
        e.printStackTrace();
        return false;
    }
    }
}

File: Note Taking Application/backend/src/dao/notedao.js
import db from "../config/db.js";
import Note from "../models/note.js";
// Get all notes for a specific user
const getNotesByUserId = (userId) => {
  return new Promise((resolve, reject) => {
    const sql = `
      SELECT noteId, title, content, dateCreated, modifiedDate, folderId, userId
      FROM notes
      WHERE userId = ?
      ORDER BY dateCreated DESC
    `;

    db.query(sql, [userId], (err, results) => {
      if (err) return reject(err);

      const notes = results.map(
        (row) =>
          new Note(
            row.noteId,
            row.title,
            row.content,
            row.dateCreated,
            row.modifiedDate,
            row.folderId,
            row.userId
          )
      );

      resolve(notes);
    });
  });
};

// Get notes that are not inside any folder
const getUnfolderedNotesByUserId = (userId) => {
  return new Promise((resolve, reject) => {
    const sql = `
      SELECT noteId, title, content, dateCreated, modifiedDate, folderId, userId
      FROM notes
      WHERE userId = ? AND folderId IS NULL
      ORDER BY dateCreated DESC
    `;

    db.query(sql, [userId], (err, results) => {
      if (err) return reject(err);

      const notes = results.map(
        (row) =>
          new Note(
            row.noteId,
            row.title,
            row.content,
            row.dateCreated,
            row.modifiedDate,
            row.folderId,
            row.userId
          )
      );

      resolve(notes);
    });
  });
};

// Get notes inside one folder
const getNotesByFolderId = (folderId, userId) => {
  return new Promise((resolve, reject) => {
    const sql = `
      SELECT noteId, title, content, dateCreated, modifiedDate, folderId, userId
      FROM notes
      WHERE folderId = ? AND userId = ?
      ORDER BY dateCreated DESC
    `;

    db.query(sql, [folderId, userId], (err, results) => {
      if (err) return reject(err);

      const notes = results.map(
        (row) =>
          new Note(
            row.noteId,
            row.title,
            row.content,
            row.dateCreated,
            row.modifiedDate,
            row.folderId,
            row.userId
          )
      );

      resolve(notes);
    });
  });
};

// Get one note by noteId
const getNoteById = (noteId, userId) => {
  return new Promise((resolve, reject) => {
    const sql = `
      SELECT noteId, title, content, dateCreated, modifiedDate, folderId, userId
      FROM notes
      WHERE noteId = ? AND userId = ?
    `;

    db.query(sql, [noteId, userId], (err, results) => {
      if (err) return reject(err);

      if (results.length === 0) {
        return resolve(null);
      }

      const row = results[0];
      const note = new Note(
        row.noteId,
        row.title,
        row.content,
        row.dateCreated,
        row.modifiedDate,
        row.folderId,
        row.userId
      );

      resolve(note);
    });
  });
};

// Create a new note
const createNote = (title, content, userId, folderId = null) => {
  return new Promise((resolve, reject) => {
    const sql = `
      INSERT INTO notes (title, content, dateCreated, modifiedDate, folderId, userId)
      VALUES (?, ?, NOW(), NULL, ?, ?)
    `;

    db.query(sql, [title, content, folderId, userId], (err, result) => {
      if (err) return reject(err);

      const newNote = new Note(
        result.insertId,
        title,
        content,
        new Date(),
        null,
        folderId,
        userId
      );

      resolve(newNote);
    });
  });
};

// Update an existing note
const updateNote = (noteId, title, content, userId) => {
  return new Promise((resolve, reject) => {
    const sql = `
      UPDATE notes
      SET title = ?, content = ?, modifiedDate = NOW()
      WHERE noteId = ? AND userId = ?
    `;

    db.query(sql, [title, content, noteId, userId], (err, result) => {
      if (err) return reject(err);
      resolve(result);
    });
  });
};

// Delete note by noteId
const deleteNote = (noteId, userId) => {
  return new Promise((resolve, reject) => {
    const sql = `
      DELETE FROM notes
      WHERE noteId = ? AND userId = ?
    `;

    db.query(sql, [noteId, userId], (err, result) => {
      if (err) return reject(err);
      resolve(result);
    });
  });
};

// Move note into a folder or remove it from folder
const moveNoteToFolder = (noteId, folderId = null, userId) => {
  return new Promise((resolve, reject) => {
    const sql = `
      UPDATE notes
      SET folderId = ?
      WHERE noteId = ? AND userId = ?
    `;

    db.query(sql, [folderId, noteId, userId], (err, result) => {
      if (err) return reject(err);
      resolve(result);
    });
  });
};

export default {
  getNotesByUserId,
  getUnfolderedNotesByUserId,
  getNotesByFolderId,
  getNoteById,
  createNote,
  updateNote,
  deleteNote,
  moveNoteToFolder
};